### DISCUSSION TITLE
Button Matrix Manipulation Question

### DISCUSSION THREAD

**Discussion by KyAZ73** [ReneRichterDE/ButtonMatrix#8] [Q&A] [answered]
Hello, thank you for this library.  I was wondering how difficult it would be to setup the button matrix for characters instead of numbers?  I am asking this so I can print out the letters on my keypad instead of its numerical representation.

Thank you

**Comment by ReneRichterDE** [comment]
Hi KyAZ73,

this should be a very simple task.

Just define an array of characters below the Button array declaration of the exact same size as the button array. Make sure the first button starts with index 0 and continuously counts up as given in the example below (0...8 here).

```
/** Button matrix button definitions */
Button buttons[ROWS][COLS] = {
    { (0), (1), (2) },
    { (3), (4), (5) },
    { (6), (7), (8) }
};

static const char buttonChar[sizeof(buttons)/sizeof(buttons[0][0])] = { 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I' };
```


and later in use the number as an index into your character array when a button is pressed ... like this


```
   .....

    if (matrix.update())
    {
        // There was a change in any of the buttons

        // Scan all buttons, except the last one
        for (uint16_t idx = 0; idx < numButtons; idx++)
        {
            pButton = matrix.getButton(idx);
            if (pButton->fell())
            {
                // Button has rose. The fell() method automatically resets the event so you cannot call it twice for the same occurrence of the event!
               //Serial.print("Button fell ");Serial.println(pButton->getNumber());
               Serial.print("Button fell ");Serial.println(buttonChar[pButton->getNumber()]);

            }
            else if (pButton->rose())
            {
                // Button has rose. The rose() method automatically resets the event so you cannot call it twice for the same occurrence of the event!
                //Serial.print("Button rose ");Serial.println(pButton->getNumber());
                Serial.print("Button rose ");Serial.println(buttonChar[pButton->getNumber()]);
            }
        }
    }

   .....

```

I have commented out the original statements from the example so you can easily see what has changed.

Of course you could as well use a string instead of a simple character array but it takes one byte more of memory for the trailing zero.

**Reply by KyAZ73** [reply]
Great, thank you. 

Can you tell me if I can perform a function depending on what key is read in (pressed)?  Would that be an "if getNumber() == 1" kind of thing?

**Reply by ReneRichterDE** [reply, marked as answer]
Not sure if I got you right. The example above shows if a button has been pushed down using the **fell()** method or has been released using **rose()** method. You can also you the **pressed()** method that repeatedly signal while a button is being pressed. The button number in the example above can be determined by the **idx** variable or you could also use the **getNumber()** method of the button to determine its number.

I strongly recommend looking at the examples includes with the library, you should find anything you need in there.

If this solves your problem, please mark the thread a solved.
